fish-rust/src/builtins/string/sub.rs

```diff
@@ -0,0 +1,105 @@
+use std::num::NonZeroI64;
+
+use super::{Arguments, StringError, StringSubCommand};
+use crate::builtins::shared::{
+    io_streams_t, BUILTIN_ERR_COMBO2, STATUS_CMD_ERROR, STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, L};
+use crate::wgetopt::{wopt, woption, woption_argument_t::*};
+use crate::wutil::{fish_wcstol, wgettext_fmt};
+
+#[derive(Default)]
+pub struct Sub {
+    length: Option<usize>,
+    quiet: bool,
+    start: Option<NonZeroI64>,
+    end: Option<NonZeroI64>,
+}
+
+impl StringSubCommand<'_> for Sub {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        wopt(L!("length"), required_argument, 'l'),
+        wopt(L!("start"), required_argument, 's'),
+        wopt(L!("end"), required_argument, 'e'),
+        wopt(L!("quiet"), no_argument, 'q'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":l:qs:e:");
+
+    fn parse_opt(&mut self, optarg: Option<&wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'l' => self.length = Some(fish_wcstol(optarg.unwrap())?.try_into()?),
+            's' => self.start = Some(fish_wcstol(optarg.unwrap())?.try_into()?),
+            'e' => self.end = Some(fish_wcstol(optarg.unwrap())?.try_into()?),
+            'q' => self.quiet = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        let cmd = args[0];
+        if self.length.is_some() && self.end.is_some() {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_COMBO2,
+                cmd,
+                L!("--end and --length are mutually exclusive")
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        let mut nsub = 0;
+        for (s, want_newline) in Arguments::new(args, optind, streams) {
+            let start: usize = match self.start.map(i64::from).unwrap_or_default() {
+                n @ 1.. => n as usize - 1,
+                0 => 0,
+                n => {
+                    let n = (-n) as usize;
```

Comment:
Since `usize` isn't 64 bit on 32bit systems, this lead to a test failure https://github.com/henrikhorluck/fish-shell/actions/runs/5641633152/job/15279997206, but fixed that also